### PR TITLE
feat: Migrate Pipedrive CRM from Revert to native Pipedrive APIs

### LINKED ISSUE #16797: Native Pipedrive integration
### Is your proposal related to a problem?

Yes, the Pipedrive integration in Cal.com is currently not functioning because it relies on Revert.dev, which has been deprecated and no longer maintained since September 2023. When I try to add Pipedrive via the web UI, I receive an error message stating "pipedrive client id missing."

### Describe the solution you'd like

I would like the Pipedrive integration to be updated to remove the dependency on Revert.dev. Instead, the integration should directly use Pipedrive's API, implementing OAuth2 authentication. This would involve updating the web UI to allow users to input their Pipedrive Client ID and Client Secret to establish the connection.

### Describe alternatives you've considered

I've considered manually integrating Pipedrive using custom API scripts, but this approach is not practical for most users due to its complexity. I also looked into third-party integration tools, but they may not offer the same seamless experience and could incur additional costs.

### Additional context

Maintaining a functional Pipedrive integration is crucial for users who rely on it to sync their scheduling with their CRM systems. The deprecation of Revert.dev has disrupted this workflow, impacting productivity. Updating the integration will restore this essential functionality for many users.

### Requirement/Document

- [Pipedrive API Authentication Documentation](https://pipedrive.readme.io/docs/core-api-concepts-authentication)

### PR DESCRIPTION
**Pull request by itsalam**
- Changed OAuth flow from environment varibles to user supplied on installation.
- Added apis to reflect User OAuth redirects
- Added setup page with instrutions
- Added dayjs and pipedrive package to the pipedrive package

## What does this PR do?

- Fixes #16797 (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description) 
  - No Linear Issue number submitted; N/A.

Loom Videos:
Adding the app: https://www.loom.com/share/d3cf7e794366418b945aacb6e599463e?sid=d926b675-02de-4d49-9aa7-052c49ddcac9
Editing events: https://www.loom.com/share/1358d47527cf4b6bb922bdd4cf42c7be?sid=21842087-03d9-43ab-85a0-2e525d27b3cf

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.
  - Seems difficult to test with current available methods. 

## How should this be tested?

- Are there environment variables that should be set?
  - None
- What are the minimal test data to have?
  - ngrok'd environment, Pipedrive test app/account 
- What is expected (happy path) to have (input and output)?
  -  Event should be created and reflect changes on Pipedrive.
- Any other important info that could help to test that PR

### CODE REVIEW

File: packages/app-store/pipedrive-crm/lib/util.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+export const getOAuthClientFromCredential = (credential: Pick<CredentialPayload, "key">) => {
+  const keys = credential.key as z.infer<typeof pipedriveAppKeysSchema>;
+  console.log(keys);
+
+  const oAuthConfig = new OAuth2Configuration({
+    clientId: keys.client_id,
+    clientSecret: keys.secret_key,
+    redirectUri: "https://9ba4a35f0b7b.ngrok.app/api/integrations/pipedrive-crm/callback", //`${WEBAPP_URL}/api/integrations/pipedrive-crm/callback`,
```
Comment (Praashh) [inline packages/app-store/pipedrive-crm/lib/util.ts:67]: redirect URL shouldn't be hard coded.

Comment (itsalam) [reply packages/app-store/pipedrive-crm/lib/util.ts:67]: Apologies, just needed to hardcode it to properly test it. Fixing now

File: packages/app-store/_pages/setup/index.tsx [hunk 2]
```diff
@@ -16,6 +16,7 @@ export const AppSetupMap = {
   sendgrid: dynamic(() => import("../../sendgrid/pages/setup")),
   stripe: dynamic(() => import("../../stripepayment/pages/setup")),
   paypal: dynamic(() => import("../../paypal/pages/setup")),
+  "pipedrive-crm": dynamic(() => import("../../pipedrive-crm/pages/setup")),
```
Comment (Praashh) [inline packages/app-store/_pages/setup/index.tsx:19]: let's keep this key same as other keys.
```suggestion
  pipedrive-crm: dynamic(() => import("../../pipedrive-crm/pages/setup")),
```

Comment (Praashh) [inline packages/app-store/_pages/setup/index.tsx:19]: let's keep this key same as other keys.
```suggestion
  pipedrive-crm: dynamic(() => import("../../pipedrive-crm/pages/setup")),
```

File: packages/app-store/pipedrive-crm/api/add.ts [hunk 3]
```diff
@@ -1,37 +1,50 @@
...
-  const appKeys = await getAppKeysFromSlug(appConfig.slug);
-  let client_id = "";
-  if (typeof appKeys.client_id === "string") client_id = appKeys.client_id;
-  if (!client_id) return res.status(400).json({ message: "pipedrive client id missing." });
   // Check that user is authenticated
   req.session = await getServerSession({ req, res });
   const { teamId } = req.query;
   const user = req.session?.user;
-  if (!user) {
+  if (user === undefined) {
```
Comment (Praashh) [inline packages/app-store/pipedrive-crm/api/add.ts:17]:
```suggestion
  if (!user) {
```

much readable.

File: packages/app-store/pipedrive-crm/api/callback.ts [hunk 4]
```diff
@@ -1,19 +1,54 @@
...
+      id: credentials.id,
+    },
+    data: {
+      key: {
+        ...credKeys,
+        tokens,
+        lastUpdated: new Date(),
+      },
+    },
+  });
```
Comment (Praashh) [inline packages/app-store/pipedrive-crm/api/callback.ts:51]: there should be checking for updating the credential, what if it fails?

File: packages/app-store/pipedrive-crm/api/setup.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { getServerSession } from "@calcom/features/auth/lib/getServerSession";
+
+import { getOAuthClientFromSession } from "../lib/util";
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (req.method !== "POST") return res.status(405).json({ message: "Method not allowed" });
+  // Check that user is authenticated
```
Comment (Praashh) [inline packages/app-store/pipedrive-crm/api/setup.ts:9]: don't think we need this comment here.

File: packages/app-store/_pages/setup/index.tsx (hunk 2 above)
Comment (itsalam) [reply packages/app-store/_pages/setup/index.tsx:19]: This change needs quotes becase of the dash "-" in its key, as noted by the other keys with dashes as well.

Comment (Praashh) [reply packages/app-store/_pages/setup/index.tsx:19]: yea got it

File: packages/app-store/pipedrive-crm/api/add.ts (hunk 3 above)
Comment (itsalam) [reply packages/app-store/pipedrive-crm/api/add.ts:17]: much ack'd

File: packages/app-store/pipedrive-crm/api/callback.ts (hunk 4 above)
Comment (itsalam) [reply packages/app-store/pipedrive-crm/api/callback.ts:51]: good point, ill update to catch that

**Review by Praashh — CHANGES_REQUESTED** [review]
we also need updated steps for pipedrive in `packages/app-store/pipedrive-crm/README.md` and `.env.appStore.example`. please add

**Comment by Praashh** [comment]
Hey @itsalam, thank you so much for your contribution but for now, we are going with the other one.

**Comment by rhinck** [comment]
> Hey @itsalam, thank you so much for your contribution but for now, we are going with the other one.

Hi @Praashh, is hosted cal.com using a Pipedrive integration without bugs currently? I see issues when trying to link our org:

image

**Pull request closed without merging**